Add OpenStack Router module

emonty opened the pull request:
Also deprecate old quantum_router module

emonty committed:
Add OpenStack Router module

Also deprecate old quantum_router module

Commit:
Update os_router to the latest version

This version uses the latest shade for a cleaner interface,
support for check mode, and updating an existing router.

bcoca commented:
requries rebase

emonty commented:
Let's hold on this one for a sec - I think we can make it and the other two router modules take better advantage of underlying library much better.

omgjlk commented on cloud/openstack/os_router.py:
```diff
@@ -0,0 +1,138 @@
+module: os_router
+short_description: Create or Delete routers from OpenStack
+extends_documentation_fragment: openstack
+version_added: "1.10"
```
This needs to be 2.0

omgjlk commented on cloud/openstack/os_router.py:
```diff
@@ -0,0 +1,138 @@
+    admin_state_up=True
+'''
+
+
```
Return data should be documented.

omgjlk commented on cloud/openstack/os_router.py:
```diff
@@ -0,0 +1,138 @@
+     required: true
+   admin_state_up:
+     description:
+        - Desired admin state of the created router.
```
This should be updated to reflect that it can change state of an existing router.

omgjlk commented on cloud/openstack/os_router.py:
```diff
@@ -0,0 +1,138 @@
+                                     id=router['id'])
+                else:
+                    module.exit_json(changed=False, result="success",
+                                     id=router['id'])
```
result= in these exit statements seem unnecessary. Changed is changed, success is assumed unless error.

omgjlk commented on cloud/openstack/os_router.py:
```diff
@@ -0,0 +1,138 @@
+# this is magic, see lib/ansible/module_common.py
+from ansible.module_utils.basic import *
+from ansible.module_utils.openstack import *
+main()
```
I think we want to do if **name** blah blah here for getting the main function to run.

gregdek commented:
@emonty thoughts on @j2sol proposed revisions here?

gregdek commented:
ping @emonty @shrews @dguerri @juliakreger @j2sol @rcarrillocruz

Shrews commented:
Agree whole-heartedly with @j2sol's comments. And I actually thought this one was already merged for some reason. Will see about getting it cleaned up.

Commit:
More os_router module cleanup and fixes.

Added a RETURN section, corrected version_added value, removed use of
'result' in exit_json() calls.

Shrews commented:
@j2sol Addressed all of your comments in the latest commit. Thanks!

Shrews commented:
Let's not merge this just yet. @emonty and I were discussing possibly merging this module with the modules from PRs #1052 and #1051 because, realistically, one wouldn't create a router without adding an interface and gateway. I will begin looking at that and will update here.

Commit:
Allow complete router configuration

This change allows one to completely configure a router, including
gateway and interfaces, using the latest shade (>0.13.0).

emonty commented:
lgtm - @j2sol ?

omgjlk commented:
I think this is good. I'll want to follow it up with another PR to add the ability to create the router for a specific tenant, instead of assuming the login tenant, but that shouldn't be part of the initial merge MVP.

Shrews commented:
@j2sol That sounds like something good to add later. Shade does not currently support that, mostly because the neutron API docs do not show that as being possible, though it clearly is now that I look at the neutron command line options.

emonty commented:
I'd love to have a deeper conversation about this concept and also pull 
in our keystone friend, because conceptually there's weird things going 
on here. But yes - as a follow on with a plan.

On 09/25/2015 02:26 PM, David Shrewsbury wrote:

> @j2sol https://github.com/j2sol That sounds like something good to add
> later. Shade does not currently support that, mostly because the neutron
> API docs do not show that as being possible, though it clearly is now
> that I look at the neutron command line options.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-core/pull/1048#issuecomment-143331110.

Shrews commented:
@gregdek @emonty With this newest module, we can close #1051 and #1052 since those are no longer needed. I can't close those myself.

Commit:
Deprecate older router modules.

The quantum_router_gateway.py and quantum_router_interface.py modules
are deprecated with this change.

Shrews commented:
@emonty @j2sol Added one commit to deprecate the quantum_router_interface and quantum_router_gateway modules since you both +1'd this. This lets us properly close #1051 and #1052 which had those deprecations.

Commit:
Add author to os_router

Shrews commented:
And added missing author line.  :)

emonty commented:
shipit

gregdek commented:
Thanks @emonty and @Shrews for the work here. Marking for inclusion.
